### PR TITLE
Use GBR to show when there is next step action required

### PR DESCRIPTION
**Pull request by rezkiy37**
### Details

We are going to use the GBR pattern to show user whenever its their turn to take an action on a report in the app.

Also, the PR tweaks prop types for next action.

### Fixed Issues

$ https://github.com/Expensify/App/issues/28768
PROPOSAL: N/A

### Tests

1. You need to be an admin of a paid "Control" workspace.
2. Invite an employee.
3. Set your admin account like an approver for the employee one.
4. Open a workspace chat from the employee account on ND.
5. Request money.
6. Open a workspace chat of the employee from the admin account.
7. Verify that there is GBR on a chat in the sidebar.

- [x] Verify that no errors appear in the JS console

### Offline tests

Same as **"Tests"**.

### QA Steps

Same as **"Tests"**.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/57314631/e91199c1-1e1f-4902-91b9-e5a5cdf307c3

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/57314631/8612f5b8-41ba-483a-9d4c-84eb20f75354

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/57314631/185cefaa-1dc2-4070-b808-76dc5fa21c1f

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/57314631/7ac92056-7e9f-489e-b8c5-647f6cca4812

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/57314631/a00880a4-a9bd-40fb-8685-9486dbc143fd

</details>

### CODE REVIEW

File: src/libs/SidebarUtils.ts [hunk 1]
```diff
@@ -298,6 +299,10 @@ function getOptionData(
     preferredLocale: ValueOf<typeof CONST.LOCALES>,
     policy: Policy,
     parentReportAction: ReportAction,
+    // TODO: Clarify typing
+    nextStep: {
+        requiresUserAction: boolean;
+    },
```
Comment (rezkiy37) [inline src/libs/SidebarUtils.ts:305]: @mountiny, how should I type `nextStep` for now?

**Comment by rezkiy37** [comment]
@mountiny, there is one moment.
In the [docs](https://docs.google.com/document/d/1vZtGpW_Fy-Wq_R6gvnjdmWVA86FnR42y3LtnNA--vcY/edit?pli=1#bookmark=id.rascpocug1m4) they mentioned this method - [isWaitingForIOUActionFromCurrentUser](https://github.com/Expensify/App/blob/d7ad30519d9632201bc711dc8a7f840cc0a8e153/src/libs/ReportUtils.js#L1210). However, currently we have [requiresAttentionFromCurrentUser](https://github.com/Expensify/App/blob/main/src/libs/ReportUtils.js#L1423).

File: src/libs/SidebarUtils.ts (hunk 1 above)
Comment (mountiny) [reply src/libs/SidebarUtils.ts:305]: For now just generic object

**Comment by mountiny** [comment]
Yep we renamed it since the [requiresAttentionFromCurrentUser](https://github.com/Expensify/App/blob/main/src/libs/ReportUtils.js#L1423) is where to look at

File: src/libs/SidebarUtils.ts (hunk 1 above)
Comment (rezkiy37) [reply src/libs/SidebarUtils.ts]: Done - 9b16ac3.

File: src/libs/SidebarUtils.ts [hunk 2]
```diff
@@ -298,6 +299,7 @@ function getOptionData(
     preferredLocale: ValueOf<typeof CONST.LOCALES>,
     policy: Policy,
     parentReportAction: ReportAction,
+    nextStep: Record<string, unknown>,
```
Comment (fabioh8010) [inline src/libs/SidebarUtils.ts:302]: Why not creating the type for `nextStep` instead of `Record<string, unknown>`?

**Comment by greg-schroeder** [comment]
@eVoloshchak can you take a look at this ASAP? Thanks!

**Comment by rezkiy37** [comment]
> Also, not sure if its a bug or not, but the `Approve` button only shows up if the policy is of type `corporate`. It does not show up when the policy is of type `team` although that is considered a `paid collect` policy as well.

I think it is something does not relate this PR. You can ask someone from the Expensify team to clarify.

**Comment by rezkiy37** [comment]
> The chat does not show up when in focus mode. I would expect the GBR chats to always show up even if the mode is focused. Can you confirm / fix @rezkiy37?

What do you mean under the "focus mode"?

For me it sounds like a requirement. @mountiny, can you confirm that I should integrate it?

**Comment by rezkiy37** [comment]
@puneetlath, @mountiny, there is a couple UX questions: https://github.com/Expensify/App/pull/31264#issuecomment-1815366566 and https://github.com/Expensify/App/pull/31264#issuecomment-1815427582.

**Comment by puneetlath** [comment]
>Also, not sure if its a bug or not, but the Approve button only shows up if the policy is of type corporate. It does not show up when the policy is of type team although that is considered a paid collect policy as well.

Yes, this should apply to Collect (team) and Control (corproate) policies. Looking at the code though, I don't think that relates to this PR. 

>What do you mean under the "focus mode"?
>For me it sounds like a requirement. @mountiny, can you confirm that I should integrate it?

Yes, we should make sure that any chat that has GBR shows up in the LHN in focus mode. @rezkiy37 focus mode is an option where you can make it so that only unread or GBR/RBR/pinned chats show up in the LHN. 

@waterim implemented the updated LHN logic that makes sure that chats with GBR are shown in focus mode. That logic is handled in [this function](https://github.com/Expensify/App/blob/ff4c947da6d5a710bb6c55b973284c13be306f75/src/libs/ReportUtils.js#L1448).  
image

**Comment by mountiny** [comment]
> Yes, this should apply to Collect (team) and Control (corproate) policies. Looking at the code though, I don't think that relates to this PR.

Yeah I think that this is mainly because what Marco worked on for the Approval for now was aimed at Control policy for C+

**Comment by rezkiy37** [comment]
I've added logic to show a report in the LHN - 4bdd124. However, I have a problem with the sidebar caching. Looks like it does not remember the latest reports and throws recently added with a next step. Investigating...

https://github.com/Expensify/App/assets/57314631/757f37c1-6575-43d6-9db1-86ce3193505a

**Comment by rezkiy37** [comment]
Btw, I've found a problem that the backend doesn't include `reportNextStep_` to basic restart/re-login command responses. It means, when I relogin, I don't have a next step anymore. An employee should request money again, after an admin gets a next step.


https://github.com/Expensify/App/assets/57314631/a81e5741-cbc7-4310-9005-0992c7fc3b50



cc @puneetlath, @mountiny

**Comment by rezkiy37** [comment]
Okay, I just noted it. Thank you!

**Comment by mountiny** [comment]
As noted in Slack, we are going to close this PR as this change actually is not desirable with the updated GBR approach we are taking where the attention is drawn to the parent chat in LHN and all the cases for Collect policies should actually be already covered in backend logic for the GBR. If we discover we need this logic in some other cases which are not covered, we can revisit this PR. Thanks for working on this and sorry for the outcome/ not catching this earlier

**Comment by mountiny** [comment]
@allroundexperts I see, I am sorry for this hassle, I think you can request the 25% of the original amount. In general, I am looking at these as something which balances out overtime, ie you can have some issues which are super simple and quick to test for $500 or more and then you have some which can take long or fall through. On average you get paid accordingly. I hope this works

**Pull request closed without merging**